File: python/src/algopay/storage/memory.py
"""In-memory storage backend."""

from __future__ import annotations

from copy import deepcopy
from typing import Any

from algopay.storage.base import StorageBackend, register_storage_backend


class InMemoryStorage(StorageBackend):
    def __init__(self) -> None:
        self._data: dict[str, dict[str, dict[str, Any]]] = {}

    def _ensure_collection(self, collection: str) -> dict[str, dict[str, Any]]:
        if collection not in self._data:
            self._data[collection] = {}
        return self._data[collection]

    async def save(self, collection: str, key: str, data: dict[str, Any]) -> None:
        coll = self._ensure_collection(collection)
        coll[key] = deepcopy(data)

    async def get(self, collection: str, key: str) -> dict[str, Any] | None:
        coll = self._ensure_collection(collection)
        data = coll.get(key)
        return deepcopy(data) if data else None

    async def delete(self, collection: str, key: str) -> bool:
        coll = self._ensure_collection(collection)
        if key in coll:
            del coll[key]
            return True
        return False

    async def query(
        self,
        collection: str,
        filters: dict[str, Any] | None = None,
        limit: int | None = None,
        offset: int = 0,
    ) -> list[dict[str, Any]]:
        coll = self._ensure_collection(collection)
        results = []
        for key, data in coll.items():
            if filters:
                match = all(data.get(fk) == fv for fk, fv in filters.items())
                if not match:
                    continue
            row = deepcopy(data)
            row["_key"] = key
            results.append(row)
        results = results[offset:]
        if limit is not None:
            results = results[:limit]
        return results

    async def update(self, collection: str, key: str, data: dict[str, Any]) -> bool:
        coll = self._ensure_collection(collection)
        if key not in coll:
            return False
        coll[key].update(deepcopy(data))
        return True

    async def count(self, collection: str, filters: dict[str, Any] | None = None) -> int:
        if filters:
            return len(await self.query(collection, filters))
        return len(self._ensure_collection(collection))

    async def clear(self, collection: str) -> int:
        coll = self._ensure_collection(collection)
        n = len(coll)
        coll.clear()
        return n

    async def atomic_add(self, collection: str, key: str, amount: str) -> str:
        from decimal import Decimal

        coll = self._ensure_collection(collection)
        cur = coll.get(key)
        if isinstance(cur, dict):
            current_dec = Decimal("0")
        else:
            current_dec = Decimal(str(cur)) if cur else Decimal("0")
        new_val = current_dec + Decimal(amount)
        coll[key] = str(new_val)
        return str(new_val)


register_storage_backend("memory", InMemoryStorage)

File: typescript/src/storage/memory.ts
import type { StorageBackend } from "./base.js";

export class InMemoryStorage implements StorageBackend {
  private readonly _data = new Map<string, Map<string, Record<string, unknown>>>();

  private _collection(name: string): Map<string, Record<string, unknown>> {
    let coll = this._data.get(name);
    if (!coll) {
      coll = new Map();
      this._data.set(name, coll);
    }
    return coll;
  }

  async save(collection: string, key: string, data: Record<string, unknown>): Promise<void> {
    this._collection(collection).set(key, structuredClone(data));
  }

  async get(collection: string, key: string): Promise<Record<string, unknown> | null> {
    const d = this._collection(collection).get(key);
    return d ? structuredClone(d) : null;
  }

  async delete(collection: string, key: string): Promise<boolean> {
    return this._collection(collection).delete(key);
  }

  async query(
    collection: string,
    filters?: Record<string, unknown> | null,
    limit?: number | null,
    offset = 0,
  ): Promise<Record<string, unknown>[]> {
    const coll = this._collection(collection);
    const results: Record<string, unknown>[] = [];
    for (const [key, data] of coll) {
      if (filters) {
        const match = Object.entries(filters).every(([fk, fv]) => data[fk] === fv);
        if (!match) continue;
      }
      results.push({ ...structuredClone(data), _key: key });
    }
    const sliced = results.slice(offset);
    return limit != null ? sliced.slice(0, limit) : sliced;
  }

  async update(collection: string, key: string, data: Record<string, unknown>): Promise<boolean> {
    const coll = this._collection(collection);
    const existing = coll.get(key);
    if (!existing) return false;
    Object.assign(existing, structuredClone(data));
    return true;
  }

  async count(collection: string, filters?: Record<string, unknown> | null): Promise<number> {
    if (filters) return (await this.query(collection, filters)).length;
    return this._collection(collection).size;
  }

  async clear(collection: string): Promise<number> {
    const coll = this._collection(collection);
    const n = coll.size;
    coll.clear();
    return n;
  }

  async atomicAdd(collection: string, key: string, amount: string): Promise<string> {
    const coll = this._collection(collection);
    const cur = coll.get(key);
    let current = 0;
    if (cur && typeof cur === "object" && !Array.isArray(cur)) {
      current = 0;
    } else if (typeof cur === "string") {
      current = Number(cur);
    }
    // For atomic add, store raw string value instead of object
    const newVal = current + Number(amount);
    coll.set(key, { _value: String(newVal) } as Record<string, unknown>);
    return String(newVal);
  }

  async healthCheck(): Promise<boolean> {
    return true;
  }
}
